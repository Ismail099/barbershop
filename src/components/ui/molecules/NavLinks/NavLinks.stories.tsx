import { Story, Meta } from '@storybook/react';

import { NavLinks } from './NavLinks';

export default {
  title: 'molecules/NavLinks',
  component: NavLinks
} as Meta;

export const Basic: Story = () => <NavLinks />;
