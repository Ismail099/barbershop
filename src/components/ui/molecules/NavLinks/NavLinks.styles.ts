import media from 'styled-media-query';
import styled, { css } from 'styled-components';

type ActiveProps = {
  active?: boolean;
};

export const Wrapper = styled.nav`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 100%;

  ${media.greaterThan('medium')`
    width: auto;
  `}
`;

export const Line = styled.span`
  ${({ theme }) => css`
    height: 0.3rem;
    position: absolute;
    bottom: 0;
    left: 0;
    background: ${theme.colors.primary};
    border-radius: 0.3rem;
    transition: all 0.3s linear;
    box-shadow: 0 0 1rem ${theme.colors.primary};
  `}
`;

export const WrapperUl = styled.ul`
  display: flex;
  gap: 4rem;
`;

export const List = styled.li`
  list-style: none;
`;

export const Link = styled.a<ActiveProps>`
  ${({ theme, active }) => css`
    text-decoration: none;
    font-size: ${theme.font.sizes.medium};
    color: ${active ? theme.colors.primary : theme.colors.white};
    text-shadow: ${active && `0 0 1rem ${theme.colors.primary}`};
    transition: 0.3s linear;
    cursor: pointer;

    &:hover,
    &:focus {
      color: ${theme.colors.primary};
      text-shadow: 0 0 1rem ${theme.colors.primary};
      outline: none;
    }
  `}
`;
