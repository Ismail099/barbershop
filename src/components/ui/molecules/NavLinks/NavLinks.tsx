import { useEffect, useRef, useState } from 'react';
import MediaMatch from '../MediaMatch/MediaMatch';
import * as S from './NavLinks.styles';

export type NavLinksProps = {
  children: React.ReactNode;
};

export const NavLinks = ({ children }: NavLinksProps) => {
  const [homeState, setHomeState] = useState(true);
  const [sobreState, setSobreState] = useState(false);
  const [servicosState, setServicosState] = useState(false);
  const [equipeState, setEquipeState] = useState(false);
  const [opiniaoState, setOpiniaoState] = useState(false);
  const [feedBackState, setFeedBackState] = useState(false);

  const link = useRef<HTMLAnchorElement>(null);
  const line = useRef<HTMLAnchorElement>(null);

  useEffect(() => {
    setTimeout(() => {
      line.current!.style.width = `${link.current!.getBoundingClientRect().width + 20}px`;
      line.current!.style.left = `${link.current!.offsetLeft - 10}px`;
    }, 50);
  }, [homeState, sobreState, servicosState, equipeState, opiniaoState, feedBackState]);

  return (
    <>
      <S.Wrapper>
        <MediaMatch greaterThan="medium">
          <S.WrapperUl>
            <S.List>
              {homeState ? (
                <S.Link
                  active
                  onClick={() => {
                    setHomeState(true);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                  ref={link}
                >
                  Home
                </S.Link>
              ) : (
                <S.Link
                  onClick={() => {
                    setHomeState(true);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                >
                  Home
                </S.Link>
              )}
            </S.List>
            <S.List>
              {sobreState ? (
                <S.Link
                  active
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(true);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                  ref={link}
                >
                  Sobre
                </S.Link>
              ) : (
                <S.Link
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(true);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                >
                  Sobre
                </S.Link>
              )}
            </S.List>
            <S.List>
              {servicosState ? (
                <S.Link
                  active
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(true);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                  ref={link}
                >
                  Serviços
                </S.Link>
              ) : (
                <S.Link
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(true);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                >
                  Serviços
                </S.Link>
              )}
            </S.List>
          </S.WrapperUl>
        </MediaMatch>

        {children}

        <MediaMatch greaterThan="medium">
          <S.WrapperUl>
            <S.List>
              {equipeState ? (
                <S.Link
                  active
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(true);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                  ref={link}
                >
                  Equipe
                </S.Link>
              ) : (
                <S.Link
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(true);
                    setOpiniaoState(false);
                    setFeedBackState(false);
                  }}
                >
                  Equipe
                </S.Link>
              )}
            </S.List>
            <S.List>
              {opiniaoState ? (
                <S.Link
                  active
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(true);
                    setFeedBackState(false);
                  }}
                  ref={link}
                >
                  Opinião
                </S.Link>
              ) : (
                <S.Link
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(true);
                    setFeedBackState(false);
                  }}
                >
                  Opinião
                </S.Link>
              )}
            </S.List>
            <S.List>
              {feedBackState ? (
                <S.Link
                  active
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(true);
                  }}
                  ref={link}
                >
                  FeedBack
                </S.Link>
              ) : (
                <S.Link
                  onClick={() => {
                    setHomeState(false);
                    setSobreState(false);
                    setServicosState(false);
                    setEquipeState(false);
                    setOpiniaoState(false);
                    setFeedBackState(true);
                  }}
                >
                  FeedBack
                </S.Link>
              )}
            </S.List>
          </S.WrapperUl>
        </MediaMatch>
      </S.Wrapper>
      <MediaMatch greaterThan="medium">
        <S.Line aria-label="Line" ref={line} />
      </MediaMatch>
    </>
  );
};
