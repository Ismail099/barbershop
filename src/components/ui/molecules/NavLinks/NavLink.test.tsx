import { screen, fireEvent } from '@testing-library/react';
import { renderWithTheme } from '~src/utils/tests/helpers';

import { NavLinks } from './NavLinks';

describe('<NavLinks />', () => {
  it('Deve renderizar todos os componentes dentro do componente <NavLinks />', () => {
    const { container } = renderWithTheme(
      <NavLinks paddingXLine={10} names={['Home', 'Sobre', 'Serviços', 'Equipe', 'Opinião', 'FeedBack']}>
        <h1>Logo</h1>
      </NavLinks>
    );

    expect(screen.getByText(/Home/i)).toBeInTheDocument();
    expect(screen.getByText(/Sobre/i)).toBeInTheDocument();
    expect(screen.getByText(/Serviços/i)).toBeInTheDocument();

    expect(screen.getByRole('heading', { name: /Logo/i })).toBeInTheDocument();

    expect(screen.getByText(/Equipe/i)).toBeInTheDocument();
    expect(screen.getByText(/Opinião/i)).toBeInTheDocument();
    expect(screen.getByText(/FeedBack/i)).toBeInTheDocument();

    expect(screen.getByLabelText(/Line/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Verificar hover do menu do componente <NavLinks />', () => {
    renderWithTheme(
      <NavLinks paddingXLine={10} names={['Home', 'Sobre', 'Serviços', 'Equipe', 'Opinião', 'FeedBack']}>
        <h1>Logo</h1>
      </NavLinks>
    );

    //Por Padrao ativado HOME
    expect(screen.getByText(/Home/i)).toHaveStyle({
      color: 'rgb(247, 24, 18)'
    });

    //Clicar no Sobre
    fireEvent.click(screen.getByText(/Sobre/i));

    expect(screen.getByText(/Home/i)).toHaveStyle({
      color: 'rgb(255, 255, 255)'
    });

    expect(screen.getByText(/Sobre/i)).toHaveStyle({
      color: 'rgb(247, 24, 18)'
    });

    //Clicar no Serviços
    fireEvent.click(screen.getByText(/Serviços/i));

    expect(screen.getByText(/Sobre/i)).toHaveStyle({
      color: 'rgb(255, 255, 255)'
    });

    expect(screen.getByText(/Serviços/i)).toHaveStyle({
      color: 'rgb(247, 24, 18)'
    });

    //Clicar no Equipe
    fireEvent.click(screen.getByText(/Equipe/i));

    expect(screen.getByText(/Serviços/i)).toHaveStyle({
      color: 'rgb(255, 255, 255)'
    });

    expect(screen.getByText(/Equipe/i)).toHaveStyle({
      color: 'rgb(247, 24, 18)'
    });

    //Clicar no Opinião
    fireEvent.click(screen.getByText(/Opinião/i));

    expect(screen.getByText(/Equipe/i)).toHaveStyle({
      color: 'rgb(255, 255, 255)'
    });

    expect(screen.getByText(/Opinião/i)).toHaveStyle({
      color: 'rgb(247, 24, 18)'
    });

    //Clicar no FeedBack
    fireEvent.click(screen.getByText(/FeedBack/i));

    expect(screen.getByText(/Opinião/i)).toHaveStyle({
      color: 'rgb(255, 255, 255)'
    });

    expect(screen.getByText(/FeedBack/i)).toHaveStyle({
      color: 'rgb(247, 24, 18)'
    });
  });
});
