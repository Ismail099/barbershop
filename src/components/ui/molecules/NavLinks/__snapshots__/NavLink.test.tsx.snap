// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavLinks /> Deve renderizar todos os componentes dentro do componente <NavLinks /> 1`] = `
.c1 {
  display: none;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 100%;
}

.c6 {
  height: 0.3rem;
  position: absolute;
  bottom: 0;
  left: 0;
  background: #F71812;
  border-radius: 0.3rem;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  box-shadow: 0 0 1rem #F71812;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 4rem;
}

.c3 {
  list-style: none;
}

.c4 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.6rem;
  color: #F71812;
  text-shadow: 0 0 1rem #F71812;
  -webkit-transition: 0.3s linear;
  transition: 0.3s linear;
  cursor: pointer;
}

.c4:hover,
.c4:focus {
  color: #F71812;
  text-shadow: 0 0 1rem #F71812;
  outline: none;
}

.c5 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.6rem;
  color: #FFFFFF;
  -webkit-transition: 0.3s linear;
  transition: 0.3s linear;
  cursor: pointer;
}

.c5:hover,
.c5:focus {
  color: #F71812;
  text-shadow: 0 0 1rem #F71812;
  outline: none;
}

@media (min-width:768px) {
  .c1 {
    display: block;
  }
}

@media (min-width:768px) {
  .c0 {
    width: auto;
  }
}

<nav
  class="c0"
>
  <div
    class="c1"
  >
    <ul
      class="c2"
    >
      <li
        class="c3"
      >
        <a
          aria-selected="true"
          class="c4"
          href="#"
        >
          Home
        </a>
      </li>
      <li
        class="c3"
      >
        <a
          aria-selected="false"
          class="c5"
          href="#"
        >
          Sobre
        </a>
      </li>
      <li
        class="c3"
      >
        <a
          aria-selected="false"
          class="c5"
          href="#"
        >
          Serviços
        </a>
      </li>
    </ul>
  </div>
  <h1>
    Logo
  </h1>
  <div
    class="c1"
  >
    <ul
      class="c2"
    >
      <li
        class="c3"
      >
        <a
          aria-selected="false"
          class="c5"
          href="#"
        >
          Equipe
        </a>
      </li>
      <li
        class="c3"
      >
        <a
          aria-selected="false"
          class="c5"
          href="#"
        >
          Opinião
        </a>
      </li>
      <li
        class="c3"
      >
        <a
          aria-selected="false"
          class="c5"
          href="#"
        >
          FeedBack
        </a>
      </li>
    </ul>
    <span
      aria-label="Line"
      class="c6"
      style="width: 20px; transform: translateX(-10px);"
    />
  </div>
</nav>
`;
